### PR TITLE
Respect the force flag on refspecs in git_remote_fetch

### LINKED ISSUE #2957: libgit2 ignores "force" refspec parameter when fetching
I can only assume Git CLT does support `+` in a fetch refspec since that's used in the default remote refspec. However but libgit2 in `git_remote_update_tips()` and specifically in `update_tips_for_spec()` does not check the force parameter of the refspec and always force-updates the ref.

The fix is pretty easy but before sending a PR, I want to make sure this behavior is not done on-purpose.

### PR DESCRIPTION
**Pull request by alexjg**
This fixes #2957 

If the `->force` flag on a refspec is not set then we check that the new refs are descendants of the current refs before updating them.

### CODE REVIEW

**Comment by ethomson** [comment]
@carlosmn any chance you can spare some 👀 for this?  Your knowledge of the networking code is far superior to mine.

**Comment by carlosmn** [comment]
Oops I guess we've been behaving like everything is forced forever. I let the tests run and it looks like there's memory errors here where one of the variables isn't initialized. Running it locally it keeps failing at 

    remote::fetch::do_update_refs_if_not_descendant_and_force [/home/carlos/git/libgit2/tests/remote/fetch.c:75]

While the overall approach seems like it should work, there do seem to be some issues in the code that can be caught by the tests.

**Comment by carlosmn** [comment]
I suspect the error happens in the case where we first create the remote-tracking branch so we look it up but `old` is left unwritten-to. Making the graph check conditional on `!error` should solve this.

**Comment by alexjg** [comment]
I notice that the test I wrote is also failing 

```
 1:   1) Failure:
1: remote::fetch::do_update_refs_if_not_descendant_and_force [../tests/remote/fetch.c:76]
1:   Function call failed: (git_commit_create(commit1id, repo1, REPO1_REFNAME, sig, sig, ((void *)0), "one", empty_tree, 0, ((void *)0)))
1:   error -15 - failed to create commit: current tip is not the first parent
```

This is strange as it passes on my machine. It looks like it's failing to clean up after the first test and so the test repo is in an unexpected state. Am I missing something on the cleanup?

**Comment by carlosmn** [comment]
It is most likely something to do with cleanup or the interaction between tests. Running just that failing test does pass.

**Comment by carlosmn** [comment]
When you create the new root commit the second time around, the branch has commits, so the empty parent list means a zero-commit as first parent, whereas the existing tip of the branch is whatever actual commit you created.

**Comment by carlosmn** [comment]
This might be the result of #5842 which is somewhat recent and might explain why you don't see it yourself if you branched off of something earlier.

**Comment by alexjg** [comment]
Ah interseting, I've rebased off `main` and now I get a test failure. Will see what I can do.

**Comment by alexjg** [comment]
I had to expose a `cl_fs_rm` function to directly remove the test repo in the cleanup code. The tests pass now but this may not be the idiomatic way to remote the repos?

**Comment by alexjg** [comment]
Hey peeps, any chance someone could re-run the tests for this?

**Comment by carlosmn** [comment]
Exposing internal clar functionality is not the way to go. Check out how other tests, e.g. the `mkdir` tests, deal with custom cleanup https://github.com/libgit2/libgit2/blob/868f4bcb4d3290f4b5320f030fccdf1e7fc8ac8a/tests/core/mkdir.c#L6-L22

**Comment by alexjg** [comment]
Thanks for the hint! I've reowrked this to use `git_futils_rmdir_r`

**Comment by alexjg** [comment]
Hey folks, sorry to bother, any chance someone can run the tests again?

**Comment by tiennou** [comment]
Sure, CI has been kicked into action.

**Comment by alexjg** [comment]
I managed to reproduce the memory sanitizer failure in a local version of the CI pipeline and I've fixed the failure there. This should be good to go now.

**Comment by alexjg** [comment]
Could someone nudge CI into action again?

**Comment by alexjg** [comment]
Looks like all the tests are passing, is there anything else to do to merge this?

**Comment by ethomson** [comment]
@carlosmn is this looking good?  :shipit:

File: tests/remote/fetch.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+
+void test_remote_fetch__initialize(void) {
+	git_config *c;
+	git_buf repo1_path_buf = GIT_BUF_INIT;
+	git_buf repo2_path_buf = GIT_BUF_INIT;
+	const char *sandbox = clar_sandbox_path();
+
+	cl_git_pass(git_buf_join(&repo1_path_buf, '/', sandbox, "fetchtest_repo1"));
```
Comment (carlosmn) [inline tests/remote/fetch.c:35]: `git_buf_joinpath` would be nicer here; it's what I was expecting and made me think you were trying to prepend slash :/

**Review by carlosmn — APPROVED** [review]
The code itself seems fine, mostly just the tests should be closer to the patterns and utilities that we already use elsewhere.

The test functions themselves could be made smaller and the "time-traveling fetch" function could also perform the check but that's minor.

**Comment by carlosmn** [comment]
One bigger change and consequence of this is that the `update_tips` callback should also be called when it doesn't get updated so it's able to report that to the user, as I believe git also does.

File: tests/remote/fetch.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	if (path[strlen(path) - 1] == '/') {
+		char* result = (char *) calloc(strlen(path) - 1, sizeof(char));
+		memcpy(result, path, strlen(path) - 1);
+		return result;
+	} else {
+		char* result = (char *) calloc(strlen(path), sizeof(char));
+		strncpy(result, path, strlen(path));
+		return result;
+	}
+}
```
[the remark this answers is not in the document]
Comment (alexjg) [reply tests/remote/fetch.c]: :facepalm: removed

File: tests/remote/fetch.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+void test_remote_fetch__initialize(void) {
+	git_config *c;
+	git_buf repo1_path_buf = GIT_BUF_INIT;
+	git_buf repo2_path_buf = GIT_BUF_INIT;
+	const char *sandbox = clar_sandbox_path();
+
+	cl_git_pass(git_buf_join(&repo1_path_buf, '/', sandbox, "fetchtest_repo1"));
+	repo1_path = calloc(repo1_path_buf.size, sizeof(char));
+	git_buf_copy_cstr(repo1_path, repo1_path_buf.size, &repo1_path_buf);
```
[the remark this answers is not in the document]
Comment (alexjg) [reply tests/remote/fetch.c]: Done

File: tests/remote/fetch.c (hunk 1 above)
Comment (alexjg) [reply tests/remote/fetch.c]: Done

**Comment by alexjg** [comment]
@carlosmn  Thanks for the review, I believe I've addressed your comments. Regarding the larger change to `update_tips`, that would imply an extension of the callback interface which could be done in a subsequent PR right?

**Comment by carlosmn** [comment]
> Regarding the larger change to `update_tips`, that would imply an extension of the callback interface which could be done in a subsequent PR right?

Yes, that would require increasing the version of the remote callback structure and adding some sort of `update_tips_ext` to include whether the update succeeded or not and why.

**Review by carlosmn — APPROVED** [review]

**Comment by ethomson** [comment]
:shipit:  Thanks for the change - and the great test @alexjg - and for the code review @carlosmn.

**Pull request merged**